scripts/updateModHistory.php

```diff
@@ -0,0 +1,219 @@
+        || $exitCode > 0) {
+		echo "failed\n";
+        exit(1);
+    }
+    $commonAncestorHash = implode("", $commonAncestor);
+    echo "done: ";
+} else {
+    echo timeStamp() . " - Last commit hash not found. Using empty git tree hash: ";
+    // since there is no modification history, generate it for all commits since the inital one
+    // 4b825dc642cb6eb9a060e54bf8d69288fbee4904 is the SHA1 of the empty git tree
+    $commonAncestorHash = "4b825dc642cb6eb9a060e54bf8d69288fbee4904";
+}
+echo $commonAncestorHash . "\n";
+
+echo timeStamp() . " - Retrieving number of files with a diff... ";
+$cmd = "git diff --name-only $commonAncestorHash $head | wc -l";
+if (exec($cmd, $numOfFilesWithDiff, $exitCode) === false
+    || $exitCode > 0) {
+    echo "failed\n";
+    exit(1);
+}
+$numOfFilesWithDiff = implode("", $numOfFilesWithDiff);
+echo "done (" . $numOfFilesWithDiff . ")\n";
+
+if ($numOfFilesWithDiff === "0") {
+    echo timeStamp() . " - No changes since last commit. Exiting...\n";
+    exit(0);
+}
+
+$modifiedFilescommand = <<<COMMAND
+#!/usr/bin/env bash
+echo "last commit hash:"
+echo "$(git rev-parse $head)"
+git diff --name-only $commonAncestorHash $head | while read -r filename; do
+  echo "filename:"
+  echo "\$filename"
+  echo "modified:"
+  echo "$(git log -1 --format='%aI' -- \$filename)"
+  echo "contributors:"
+  git log --format='%an' -- \$filename|awk '!a[$0]++'
+done
+COMMAND;
+
+echo timeStamp() . " - Retrieving commit authors and last commit date/time of modified files... \n";
+
+$modifiedFiles = [];
+
+$proc = popen($modifiedFilescommand, 'rb');
+while (($line = fgets($proc)) !== false) {
+    processGitDiffLine(rtrim($line, "\n\r"), $modifiedFiles);
+    if (! $runningInGithubActions) {
+        $fileCounter = max(count($modifiedFiles) - 1, 0);
+        fwrite(
+            STDERR,
+            sprintf("\033[0G{$fileCounter} of {$numOfFilesWithDiff} files read...", "", "")
+        );
+    }
+}
+pclose($proc);
+
+echo " done\n";
+
+$s = ($modifiedFiles > 2) ? "s" : "";
+echo timeStamp() . " - Retrieved author$s and last commit date$s/time$s for " . (count($modifiedFiles) - 1) . " file$s\n";
+if (count($modifiedFiles) === 1) {
+    // there will always be at least 1 entry with the last commit hash
+    exit(1);
+}
+
+$mergedModHistory = array_merge($modHistoryArray, $modifiedFiles);
+
+echo timeStamp() . " - Writing modification history file... ";
+
+$fp = fopen($modHistoryFile, "w");
+fwrite($fp, "<?php\n\n/* This is a generated file */\n\nreturn [\n");
+foreach ($mergedModHistory as $fileName => $fileProps) {
+    if ($fileName === "last commit hash") {
+        fwrite($fp, "    \"last commit hash\" => \"" . implode("", $fileProps) . "\",\n");
+        continue;
+    }
+    $newModHistoryString = '    "' . $fileName . "\" => [\n";
+    $newModHistoryString .= "        \"modified\" => \"" . ($fileProps["modified"] ?? "") . "\",\n";
+    $newModHistoryString .= "        \"contributors\" => [\n";
```

Comment:
Use sprintf or var_export here?

Reply:
Something like this?

```php
$newModHistoryString = sprintf("%4s\"%s\" => [\n%8s\"modified\" => \"%s\",\n%8s\"contributors\" => [\n", 
    " ", 
    $fileName, 
    " ", 
    $fileProps["modified"] ?? "",
    " "
);
```

Reply:
Possibly, but this really minor actually.